python/dgl/runtime/scheduler.py

```diff
@@ -746,26 +774,36 @@ def _gen_send_reduce(
     var_send_edges : var.IDX
         The edges (ids) to perform send.
     var_reduce_nodes : var.IDX
-        The nodes to perform reduce. This should include unique(v) + 0deg nodes.
+        Unique and sorted nodes to perform reduce. This should include
+        unique(v) + 0deg nodes.
     uv_getter : callable
-        A function that returns a pair of var.IDX (u, v) for the triggered edges.
+        Function that returns a pair of var.IDX (u, v) for the triggered edges.
     adj_creator : callable
-        A function that returns the adjmat and the shuffle index.
-    inc_creator : callable
-        A function that returns the incmat and the shuffle index.
+        Function that returns the adjmat, edge order of csr matrix, and edge
+        order of transposed csr matrix
+    out_map_creator: callable
+        A function that returns a mapping from reduce_nodes to relabeled
+        consecutive ids
```

Comment:
Why do we need this as an argument?

Reply:
Think about performing recv on only a few nodes. If we use UDF reducer, even if these recv nodes are not consecutive (let's say node 5, 10, 20, 100), the output will just be a compact tensor of 4 rows. You can think of this as an implicit mapping from global id 5, 10, 20, 100 to local id 0, 1, 2, 3. 

Builtin function should result in the same behavior as UDF. But kernels think the entire graph. Therefore, we need a way to tell kernel that it needs to write the output to some local id space 0, 1, 2, 3 instead of the global id space. That's why we need the output map.

The output map is different for different APIs. Update_all needs all nodes, so local space == global space. But send_and_recv needs this out mapping. To unify, we have this out_map_creator and use it to create out_map lazily.